proposals/NNNN-task-executor-preference.md

```diff
@@ -0,0 +1,825 @@
+
+The new protocol mirrors `Executor` and `SerialExecutor` in API, however it provides different semantics, and is tracked using a different mechanism at runtime -- by obtaining it from a task's executor preference record.
+
+The `TaskExecutor` is defined as:
+
+```swift
+public protocol TaskExecutor: Executor {
+  func enqueue(_ job: consuming ExecutorJob)
+
+  func asUnownedTaskExecutor() -> UnownedTaskExecutor
+}
+```
+
+As an intuitive way to think about `TaskExecutor` and `SerialExecutor`, one can think of the prior as being a "source of threads" to execute work on,
+and the latter being something that "provides serial isolation" and is a crucial part of Swift actors. The two share similarities, however the task executor has a more varied application space.
+
+### Task executor preference inheritance in Structured Concurrency
+
+Task executor preference is inherited by child tasks and actors which do not declare an explicit executor (so-called "default actors"), and is *not* inherited by un-structured tasks. 
+
+Specifically:
+
+* **Do** inherit task executor preference
+    * TaskGroup’s `addTask()`, unless overridden with explicit parameter
+    * `async let`
+    * methods on default actors (actors which do not use a custom executor)
+* **Do not** inherit task executor preference
+    * Unstructured tasks: `Task {}` and `Task.detached {}`
+    * methods on actors which **do** use a custom executor (including e.g. the `MainActor`)
+
+This also means that an entire tree can be made to execute their nonisolated work on a specific executor, just by means of setting the preference on the top-level task.
+
+### Task executor preference and async let
+
+Since `async let` are the simplest form of structured concurrency, they dot not offer much in the way of customization.
+
+An async currently always executes on the global concurrent executor, and with the inclusion of this proposal, it does take into account task executor preference. In other words, if an executor preference is set, it will be used by async let to enqueue its underlying task:
+
+```swift
+func test() async -> Int {
+  return 42
+}
+
+await withTaskExecutor(someExecutor) { 
+  async let value = test() // async let's "body" and target function execute on 'someExecutor'
+  // ... 
+  await value
+}
+```
+
+### Task executor preference and TaskGroups
+
+A `TaskGroup` and its various friends (`ThrowingTaskGroup`, `DiscardingTaskGroup`, ...) are the most powerful, but also most explicit and verbose API for structured concurrency. A group allows creating multiple child tasks using the `addTask` method, and always awaits all child tasks to complete before returning.
+
+This proposal adds overloads to the `addTask` method, which changes the executor the child tasks will be enqueued on:
+
+```swift
+extension (Discarding)(Throwing)TaskGroup { 
+  mutating func addTask(
+    on executor:  (any TaskExecutor)?,  // 'nil' means the global pool
```

Comment:
Okay, this is non-editorial, but I'd rather we avoid existentials in the interface if we can. 
```suggestion
    on executor:  (some TaskExecutor)?,  // 'nil' means the global pool
```

Reply:
This is subtle, here's the difference:

```
addTask() { ... } // no preference
addTask(on: nil) { ... } // global pool
addTask(on: something) { ... } // specific pool
```

Alternatively we could introduce a three-value enum:

```
on: .taskExecutor(...)
on: .noPreference
on: .defaultGlobalExecutor
```

?

Reply:
These are `any` on purpose but I could revisit this -- I think this is just because I struggled with implementation to get the existential that we need to store hmmm... 

We're not allowing getting the current executor which would have been `any TaskExecutor` so that necessity to make this any as well is gone...

Think we should try to make these `some`?

Reply:
I'd like to revisit this, but it can be part of a review

Reply:
Yup, happy to do so!

Reply:
Personally, I still thing using `nil` for global pool is wrong and an explicit enums makes more sense.

Reply:
Sure, happy to do very explicit enums -- I quite like them but no APIs in Swift really do it so far, we can though 👍